### PR TITLE
feat(destination): Add http2 client params to endpoints

### PR DESCRIPTION
**Pull request by olix0r**
This change updates the Destination API so that its WeightedAddr type, which represents an individual network endpoint, with HTTP/2 client parameters. This enables the control plane to alter keep-alive and flow control settings on a per-endpoint basis.

### CODE REVIEW

File: proto/destination.proto [hunk 1]
```diff
@@ -166,6 +172,47 @@ message ProtocolHint {
...
+  message KeepAlive {
+    // The time between pings.
+    google.protobuf.Duration interval = 1;
+
+    // The time to wait for a ping response before considering the connection
+    // dead.
+    google.protobuf.Duration timeout = 2;
+
+    // Whether to send pings when there is no other traffic.
+    bool while_idle = 3;
```
Comment (adleong) [inline proto/destination.proto:206]: what are the semantics of this? if false will pings be send when there IS other traffic? if true will pings ONLY be sent when there is no other traffic?

File: proto/destination.proto [hunk 2]
```diff
@@ -166,6 +172,47 @@ message ProtocolHint {
   }
 }
 
+// Configures the parameters used to initialize an HTTP/2 connection.
+message Http2ClientParams {
+  // Overrides the default client flow control settings.
+  FlowControl flow_control = 1;
+
+  // Enables keep-alive timeouts.
+  KeepAlive keep_alive = 2;
```
Comment (adleong) [inline proto/destination.proto:181]: Can keep-alives be disabled entirely? is that what the `while_idle` field does?

File: proto/destination.proto [hunk 3]
```diff
@@ -166,6 +172,47 @@ message ProtocolHint {
...
+
+  // Configures Hyper internals.
+  Internals internals = 3;
+
+  message FlowControl {
```
Comment (adleong) [inline proto/destination.proto:186]: is the idea that the control plane will always provide these values? or can the control plane leave them unspecified and allow the proxy to use some defaults?

File: proto/destination.proto (hunk 1 above)
Comment (olix0r) [reply proto/destination.proto:206]: From [hyper](https://docs.rs/hyper/latest/hyper/client/conn/http2/struct.Builder.html#method.keep_alive_while_idle)
> Sets whether HTTP2 keep-alive should apply while the connection is idle.
>
> If disabled, keep-alive pings are only sent while there are open request/responses streams. If enabled, pings are also sent when no streams are active. Does nothing if keep_alive_interval is disabled.

File: proto/destination.proto (hunk 3 above)
Comment (olix0r) [reply proto/destination.proto:186]: If the control plane omits them, the proxy will use defaults. I don't have immediate plans to set these values from the control plane.

**Comment by olix0r** [comment]
Related: https://github.com/linkerd/linkerd2/pull/12504 https://github.com/linkerd/linkerd2-proxy/compare/ver/h2-keepalives-endpoint-api

**Review by adleong — APPROVED** [review]

**Review by alpeb — APPROVED** [review]

**Pull request merged**